src/Configuration/Programmer/Programmer_helper.cpp

```diff
@@ -0,0 +1,453 @@
+    while (lineStream >> token) {
+      tokens.push_back(token);
+    }
+    if (tokens.size() == 8) {
+      TapInfo tapInfo;
+      tapInfo.index = CFG_convert_string_to_u64(tokens[0]);
+      tapInfo.tapName = tokens[1];
+      tapInfo.enabled = tokens[2] == "Y" ? true : false;
+      tapInfo.idCode =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[3], 0, 0));
+      tapInfo.expected =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[4], 0, 0));
+      tapInfo.irLen =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[5]));
+      tapInfo.irCap =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[6], 0, 0));
+      tapInfo.irMask =
+          static_cast<uint32_t>(CFG_convert_string_to_u64(tokens[7], 0, 0));
+      tapInfoList.push_back(tapInfo);
+    }
+  }
+
+  return tapInfoList;
+}
+/*
+This function expecting the string from openOCD gemini list output
+with example below
+
+         Device               ID           IRLen      Flash Size
+-------- -------------------- ------------ ---------- ----------------
+Found  0 Gemini               0x20000913   5          16384
+*/
+int extractDeviceList(const std::string& devicesString,
+                      std::vector<Device>& devices) {
+  devices.clear();
+
+  std::istringstream iss(devicesString);
+  std::string line;
+  size_t pos = std::string::npos;
+  std::vector<std::string> matches;
+  const std::string pattern(
+      R"(Found\s+(\d+)\s+([\w\s]+)\s+([0-9a-fA-Fx]+)\s+(\d+)\s+(\d+))");
+
+  while (std::getline(iss, line)) {
+    matches = findStringPattern(line, pattern);
+    if (matches.size() == 0) {
+      continue;
+    }
+    // Replace all the spaces with a single space
+    while ((pos = line.find("  ")) != std::string::npos) {
+      line.replace(pos, 2, " ");
+    }
+    Device device;
+    std::istringstream lineStream(line);
+    std::string firstToken;
+    std::getline(lineStream, firstToken, ' ');  // "found string"
+
+    std::string indexStr;
+    std::getline(lineStream, indexStr, ' ');  // index string
+    device.index = static_cast<int>(CFG_convert_string_to_u64(indexStr));
+
+    std::getline(lineStream >> std::ws, device.name, ' ');  // device name
+
+    std::string idCodeStr;
+    std::getline(lineStream >> std::ws, idCodeStr,
+                 ' ');  // idCode string in hex
+    device.tapInfo.idCode =
+        static_cast<uint32_t>(CFG_convert_string_to_u64(idCodeStr));
+
+    std::string irLenStr;
+    std::getline(lineStream >> std::ws, irLenStr, ' ');  // irlength string
+    device.tapInfo.irLen =
+        static_cast<uint32_t>(CFG_convert_string_to_u64(irLenStr));
+
+    std::string flashSizeStr;
+    std::getline(lineStream >> std::ws, flashSizeStr,
+                 ' ');  // flash size string in base 10
+    device.flashSize =
+        static_cast<uint32_t>(CFG_convert_string_to_u64(flashSizeStr));
+    // check flashSize is multiple of 1024
+    if (device.flashSize % 1024 != 0) {
```

Comment:
As I mentioned in original comment, the checking of flash size should be
1. multiple of 8 (because it is in bits, we do not have 6bits memory for example) 
2. powerof2 which mean it should be 8, 16, 32, 64, 128, 1024, 2048, 4096, 8k, 16k, 32k, 1M, 2M, 4M, 8M, 128M, 1G, 512G etc

If the flashsize is 3*1024 = 3072, which can be divided by 1024 but it is still an invalid size. I do not think you can buy embedded flash with 7M, 800M size.

Reply:
Fixed.